### PR TITLE
fix(gr00t_inference): symlink-resolve volume guard + posttrain docstring split (#391)

### LINKED ISSUE #213: docs(gr00t_inference): symmetric posttrain disclaimer for all POSTTRAIN_TAGS embodiments
## Context

Follow-up from PR #128 review (R3, R5, R7 threads).

The `gr00t_inference` tool docstring now warns that `unitree_g1_sonic` requires a finetuned checkpoint, but per upstream `Isaac-GR00T@3df8b38/gr00t/data/embodiment_tags.py::POSTTRAIN_TAGS`, the full posttrain set is:

```
POSTTRAIN_TAGS = {
    EmbodimentTag.UNITREE_G1,           # -> unitree_g1_full_body_with_waist_height_nav_cmd
    EmbodimentTag.UNITREE_G1_SONIC,     # -> unitree_g1_sonic         (covered by #128)
    EmbodimentTag.SIMPLER_ENV_GOOGLE,
    EmbodimentTag.SIMPLER_ENV_WIDOWX,
    EmbodimentTag.LIBERO_PANDA,         # -> libero_sim
}
```

## Problem

A user reading the `gr00t_inference` docstring today would assume all listed Unitree-G1 / LIBERO embodiments are interchangeable with `unitree_g1_real` (which is in `PRETRAIN_TAGS`, baked into `nvidia/GR00T-N1.7-3B`). They aren't -- pointing the base checkpoint at any posttrain tag silently emits garbage actions.

## Additional concern raised in PR #128 R7 review

The disclaimer says `requires a finetuned checkpoint`, which warns about the **model side** but not the **action side**: `action.motion_token` returned by SONIC inference is a 64-dim learned latent, NOT an executable joint command. A user wiring the policy output straight into a Unitree G1 SDK call gets garbage even with a properly fine-tuned checkpoint, because the SONIC C++ decoder (separate process, 50 Hz, from `NVlabs/GR00T-WholeBodyControl`) is in the loop.

Reviewer suggestion (PR #128 R7, non-blocking):

> ``unitree_g1_sonic`` (SONIC whole-body controller -- VLA emits 64-dim motion-token latents that must be decoded to joint commands by the SONIC runtime from GR00T-WholeBodyControl; requires a finetuned checkpoint, not the base nvidia/GR00T-N1.7-3B)

## Proposal

Replace the per-line disclaimer (currently only on the SONIC entry) with a structural split that mirrors upstream's `PRETRAIN_TAGS` / `POSTTRAIN_TAGS` partitioning, AND extend the SONIC entry to call out the decoder requirement. Sketch:

```
**Pretrain configs (work directly with nvidia/GR00T-N1.7-3B):**
  - ``unitree_g1_real``
  - ``fourier_gr1_arms_only`` ...

**Posttrain configs (require a finetuned checkpoint, not the base model):**
  - ``unitree_g1`` / ``unitree_g1_full_body``
  - ``unitree_g1_sonic`` (action stream is SONIC motion-token latents;
    requires GR00T-WholeBodyControl decoder to produce joint commands)
  - ``libero_panda`` / ``libero_sim``
  - ``simpler_env_google``, ``simpler_env_widowx``
```

This scales as upstream adds new posttrain entries: one footnote per category instead of per-line annotations.

## Pin contract

Update / extend `tests/policies/groot/test_data_config.py::test_gr00t_inference_docstring_flags_sonic_as_posttrain` into a parametrised test covering every entry in upstream's `POSTTRAIN_TAGS` so a future addition of a posttrain embodiment fails CI until the docstring is updated.

Additionally, harden the SONIC-specific pin so EVERY occurrence of `` ``unitree_g1_sonic`` `` in the docstring carries the disclaimer, not just the first (R7 reviewer suggestion):

```python
import re
occurrences = re.findall(r"``unitree_g1_sonic``[^`]*", doc)
assert occurrences
for occ in occurrences:
    assert "finetuned checkpoint" in occ, f"missing disclaimer in: {occ!r}"
```

## References

- PR #128 (R3): added the SONIC-only disclaimer
- PR #128 (R5): flagged asymmetry with `unitree_g1_full_body` and `libero_panda`
- PR #128 (R7): flagged motion_token decoder gap and pin-test enumeration gap
- Upstream: `Isaac-GR00T@3df8b38/gr00t/data/embodiment_tags.py`
- Upstream decoder: `NVlabs/GR00T-WholeBodyControl`

### PR DESCRIPTION
**Pull request by cagataycali**
## Summary

`gr00t_inference` hardening + docs bundle for the v0.4.0 umbrella (#391).

## Changes

| Issue | Fix | Kind |
|-------|-----|------|
| **#384** item 2 | Volume-safety guard ran `normpath` only — a pre-existing host symlink pointing at a protected dir (`/data/ckpt -> /etc`) passed the blocklist while docker mounted the resolved target. Added `_resolve_host_path` (realpath) and now check **both** the normalized and symlink-resolved forms against the blocklist. Residual check-vs-mount-time TOCTOU documented. | prod |
| **#384** item 1 | Documented the wholesale `/home` block decision inline (`NOTE`): it is intentional defence-in-depth; operators place checkpoints outside `/home`. The auto-derived default (`~/.cache/huggingface`) is never agent-controlled. | docs |
| **#213** | Restructured the embodiment docstring into **pretrain vs posttrain**. Marked every upstream `POSTTRAIN_TAGS` entry `[posttrain]` and added a footnote stating they require a finetuned checkpoint, not the base `nvidia/GR00T-N1.7-3B`. Extended the SONIC entry to call out that the action stream is 64-dim **motion-token latents** needing the `NVlabs/GR00T-WholeBodyControl` decoder. | docs |

## Resolved-by-refactor (closed for tracking)

`#327` (IPv6 `-p` bracketing), `#328` (`_validate_path` empty-string + `lifecycle` validator-bypass) and `#329` (`validate_inputs` rename) all targeted the **pre-#372** validation API. The agent-controlled host/volume/command surface was removed in #372 — the tool no longer threads `host` into docker `-p` (`-p {port}:{port}`), and `validate_inputs`/`_validate_path` no longer exist. These are resolved by that refactor; pinned closed here. The current agent-surface invariant remains pinned by `test_no_agent_controlled_host_exec_or_path_params`.

## Test plan

```
python -m pytest tests/tools/test_gr00t_container_hardening.py tests/policies/groot/test_data_config.py -q  # 121 passed
ruff check . && ruff format --check .   # clean
mypy strands_robots/tools/gr00t_inference.py   # Success
```

Regression pin verified: removing the realpath candidate from `_check_volume_safety` makes `test_check_volume_safety_resolves_symlink_into_protected_dir` fail; restoring it passes.

Closes #213, #327, #328, #329, #384.
Part of #391.

### CODE REVIEW

**Review by yinsong1986 — APPROVED** [review]
## Summary

Two-part hardening + docs bundle on `gr00t_inference`:

1. **#384 item 2 (security)**: `_check_volume_safety` previously normalised host paths via `normpath` only, so a pre-existing host symlink whose target was a protected dir (e.g. `/data/ckpt -> /etc`) passed the blocklist while docker mounted the resolved target. The fix adds `_resolve_host_path` (`os.path.realpath`) and now checks **both** the normalized and symlink-resolved forms against `_BLOCKED_VOLUME_EXACT` and the prefix blocklist. Residual check-vs-mount-time TOCTOU is documented honestly in the new helper's docstring.
2. **#384 item 1 + #213 (docs)**: a `NOTE` block explains the wholesale `/home` block as intentional defence-in-depth, and the embodiment docstring is restructured into pretrain vs posttrain with `[posttrain]` markers on every upstream `POSTTRAIN_TAGS` entry plus a footnote that they require a finetuned checkpoint (the SONIC entry now also calls out the 64-dim motion-token latent action stream + `NVlabs/GR00T-WholeBodyControl` decoder requirement).

Diff is +134/-8 across the tool module and two test files; no API surface changes, no wire/persisted-schema changes, no new defaults that silently invalidate older configs.

## What's good

- Realpath check is paired with a regression test (`test_check_volume_safety_resolves_symlink_into_protected_dir`) that the PR description verified fails on pre-fix code — pins the fix per AGENTS.md > Review Learnings (#86) > 'Pin every reviewed fix with a regression test'.
- Companion test (`test_check_volume_safety_allows_symlink_into_safe_dir`) confirms the guard does not over-reject legitimate symlinks into operator-chosen dirs.
- Posttrain docstring pin enumerates the upstream `POSTTRAIN_TAGS` set so a future addition lands marked or fails CI.
- `_resolve_host_path` docstring explicitly acknowledges the residual TOCTOU rather than papering over it; matches AGENTS.md > 'Operational Runbooks for Security Pins' tone.
- No host paths, no emojis, no agent-controlled surface widened.

## Verification suggestions

Standard CI is sufficient. For a quick local spot-check:

```bash
hatch run test tests/tools/test_gr00t_container_hardening.py tests/policies/groot/test_data_config.py -q
```
